### PR TITLE
feat(public-cloud): default payment method choice

### PR DESCRIPTION
**Pull request by Jisay**
| Question         | Answer
| ---------------- | ---
| Branch?          | `develop`
| Bug fix?         | no
| New feature?     | yes
| Breaking change? | no
| Tickets          | MANAGER-4345
| License          | BSD 3-Clause

## Description

The customers that have a validated payment mean but not chosen as the "by default" one see their first valid payment mean taken automatically, and considered as "by default", without any notification. This feature enable users to select an existing payment method or by creating a new one and set it as default.

## 🔖 TODOs

- [x] TRANS-26177

### CODE REVIEW

File: packages/manager/modules/pci/src/projects/new/classes/eligibility.class.js [hunk 1]
```diff
@@ -33,4 +35,12 @@ export default class PciEligibility {
   isVerifyPaypalRequired() {
     return this.actionsRequired.includes(ELIGIBILITY_ACTION_ENUM.VERIFY_PAYPAL);
   }
+
+  setValidPaymentMethods(validPaymentMethods) {
+    this.validPaymentMethods = validPaymentMethods;
+  }
+
+  isDefaultPaymentMethodChoiceRequired() {
```
Comment (Jisay) [inline packages/manager/modules/pci/src/projects/new/classes/eligibility.class.js:43]: I set it there in order to have all requirements at the same place. It simply extends a little bit the eligibility API response in order to add a new test.

File: packages/manager/modules/pci/src/projects/new/payment/components/add/component.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import controller from './controller';
```
Comment (Jisay) [inline packages/manager/modules/pci/src/projects/new/payment/components/add/component.js:1]: Use a component as payment method registration can be done in two situations : if no payment method at all or no default payment method is setted as default.

File: packages/manager/modules/pci/src/projects/new/payment/components/add/index.html [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            data-type="warning"
+            data-ng-if="$ctrl.model.paymentMethod.paymentType === 'BANK_ACCOUNT'"
+        >
+            <span
+                data-translate="pci_project_new_payment_method_add_warning_bank_account"
+            ></span>
+        </oui-message>
+
+        <p>
+            <small
```
Comment (FredericEspiau) [inline packages/manager/modules/pci/src/projects/new/payment/components/add/index.html:40]: `small` has a semantic meaning: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/small

is it he case here ?

File: packages/manager/modules/pci/src/projects/new/payment/components/add/translations/messages_FR_fr.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+  "pci_project_new_payment_method_add_explain": "L'enregistrement d'un moyen de paiement est nécessaire afin de valider votre identité. Tant que vous ne consommez rien vous ne serez pas débité ou prélevé.",
+  "pci_project_new_payment_method_add_info_credit_card": "En cliquant sur « Créer mon projet » vous allez être redirigé sur l’interface sécurisée de notre prestataire pour enregistrer une nouvelle carte bancaire. Une empreinte financière est faite auprès de votre organisme bancaire afin de valider la saisie et la validité de votre carte. Aucun montant ne sera prélevé.",
```
Comment (FredericEspiau) [inline packages/manager/modules/pci/src/projects/new/payment/components/add/translations/messages_FR_fr.json:3]:
```suggestion
  "pci_project_new_payment_method_add_info_credit_card": "En cliquant sur « Créer mon projet » vous allez être redirigé vers l’interface sécurisée de notre prestataire pour enregistrer une nouvelle carte bancaire. Une empreinte financière est faite auprès de votre organisme bancaire afin de valider la saisie et la validité de votre carte. Aucun montant ne sera prélevé.",
```
?

File: packages/manager/modules/pci/src/projects/new/payment/components/choose/index.scss [hunk 5]
```diff
@@ -0,0 +1,76 @@
+.pci-project-new-payment-choose {
```
Comment (Jisay) [inline packages/manager/modules/pci/src/projects/new/payment/components/choose/index.scss:1]: Need to override some styles from ui-kit/ods in order to fit as more as possible the mockups

File: packages/manager/modules/pci/src/projects/new/payment/components/register/index.html [hunk 6]
```diff
@@ -4,50 +4,11 @@
         data-translate="pci_project_new_payment_register_title"
     ></h2>
 
-    <div
-        class="row"
-        data-ng-repeat="authorizedPaymentMethods in $ctrl.authorizedPaymentMethods.chunks track by $index"
+    <pci-project-new-payment-method-add
```
Comment (Jisay) [inline packages/manager/modules/pci/src/projects/new/payment/components/register/index.html:7]: Centralize payment method registration in an other component

File: packages/manager/modules/pci/src/projects/new/payment/components/add/translations/messages_FR_fr.json (hunk 4 above)
Comment (Jisay) [reply packages/manager/modules/pci/src/projects/new/payment/components/add/translations/messages_FR_fr.json:3]: This content is not yet defined. I have asked the right one ;-)

File: packages/manager/modules/pci/src/projects/new/payment/components/add/index.html [hunk 7]
```diff
@@ -0,0 +1,48 @@
+<div class="pci-project-new-payment-method-add">
+    <div
+        class="row"
+        data-ng-repeat="authorizedPaymentMethods in $ctrl.authorizedPaymentMethods.chunks track by $index"
```
Comment (marie-j) [inline packages/manager/modules/pci/src/projects/new/payment/components/add/index.html:4]: The `col-md-6` isn't enough to have two elements per `row` ?

File: packages/manager/modules/pci/src/projects/new/payment/components/add/index.html [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+                data-disabled="$ctrl.globalLoading.finalize"
+            >
+                <span
+                    data-ng-bind="$ctrl.getPaymentMethodTypeText(authorizedPaymentMethod)"
+                ></span>
+            </oui-radio>
+        </div>
+    </div>
+
+    <div data-ng-if="$ctrl.model.paymentMethod.paymentType !== 'CREDIT'">
```
Comment (marie-j) [inline packages/manager/modules/pci/src/projects/new/payment/components/add/index.html:27]: Maybe some utilities function like `isCredit()`, `isBankAccount()` in the `PaymentMethod` model could be useful to avoid using constants in the view

File: packages/manager/modules/pci/src/projects/new/payment/components/add/index.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import angular from 'angular';
+import '@ovh-ux/ng-translate-async-loader';
+import 'angular-translate';
+
+import component from './component';
+
+const moduleName = 'pciProjectNewPaymentMethodAdd';
+
+angular
+  .module(moduleName, ['ngTranslateAsyncLoader', 'pascalprecht.translate'])
```
Comment (marie-j) [inline packages/manager/modules/pci/src/projects/new/payment/components/add/index.js:10]:
```suggestion
  .module(moduleName, ['oui', 'ngTranslateAsyncLoader', 'pascalprecht.translate'])
```

File: packages/manager/modules/pci/src/projects/new/payment/components/choose/index.scss [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        .oui-radio__label {
+          margin-bottom: 0;
+
+          span:nth-child(1) {
+            display: flex;
+          }
+
+          .payment-method-image-container {
+            width: 44px;
+            max-width: 44px;
```
Comment (varun257) [inline packages/manager/modules/pci/src/projects/new/payment/components/choose/index.scss:56]: Can use `rem` instead

File: packages/manager/modules/pci/src/projects/new/payment/controller.js [hunk 11]
```diff
@@ -153,9 +154,26 @@ export default class PciProjectNewPaymentCtrl {
...
+      this.globalLoading.setDefaultPaymentMehtod = true;
+
+      defaultPaymentMethodPromise = this.ovhPaymentMethod
+        .setPaymentMethodAsDefault(this.model.defaultPaymentMethod)
+        .finally(() => {
+          this.globalLoading.setDefaultPaymentMehtod = false;
+        });
+    }
+
+    return Promise.all([challengePromise, defaultPaymentMethodPromise]).then(
```
Comment (varun257) [inline packages/manager/modules/pci/src/projects/new/payment/controller.js:170]: Is there is no error handling for this ?

File: packages/manager/modules/pci/src/projects/new/payment/components/choose/index.scss (hunk 5 above)
Comment (marie-j) [reply packages/manager/modules/pci/src/projects/new/payment/components/choose/index.scss:1]: I can understand that some overriding is needed now and then, but shouldn't the latest mockup design then be integrated in ui-kit/ods and therefore the UI changes automatically applied ?

File: packages/manager/modules/pci/src/projects/new/payment/controller.js (hunk 11 above)
Comment (marie-j) [inline packages/manager/modules/pci/src/projects/new/payment/controller.js:170]: No issue with `$scope` digest ?

File: packages/manager/modules/pci/src/projects/new/payment/routing.js [hunk 12]
```diff
@@ -38,8 +50,24 @@ export default /* @ngInject */ ($stateProvider) => {
       paymentStatus: /* @ngInject */ ($transition$) =>
         $transition$.params().paymentStatus,
 
-      defaultPaymentMethod: /* @ngInject */ (ovhPaymentMethod) =>
-        ovhPaymentMethod.getDefaultPaymentMethod(),
+      paymentMethods: /* @ngInject */ (ovhPaymentMethod) =>
+        ovhPaymentMethod.getAllPaymentMethods(),
```
Comment (marie-j) [inline packages/manager/modules/pci/src/projects/new/payment/routing.js:54]: I don't know how much impact there is to get all methods but maybe we should only do so if we know it's needed (like when there is no default payment method for instance)

File: packages/manager/modules/pci/src/projects/new/routing.js [hunk 13]
```diff
@@ -152,6 +153,7 @@ export default /* @ngInject */ ($stateProvider) => {
       },
 
       globalLoading: () => ({
+        setDefaultPaymentMehtod: false,
```
Comment (marie-j) [inline packages/manager/modules/pci/src/projects/new/routing.js:156]:
```suggestion
        setDefaultPaymentMethod: false,
```
😄

File: packages/manager/modules/pci/src/projects/new/payment/index.html [hunk 14]
```diff
@@ -80,11 +80,13 @@ <h1 data-translate="pci_project_new_payment_title"></h1>
...
-                        ($ctrl.eligibility.isChallengePaymentMethodRequired() && !$ctrl.model.challenge.isValid($ctrl.defaultPaymentMethod.paymentType)
+                        $ctrl.eligibility.isChallengePaymentMethodRequired() && !$ctrl.model.challenge.isValid($ctrl.defaultPaymentMethod.paymentType)
                         || (!$ctrl.model.paymentMethod && $ctrl.eligibility.isAddPaymentMethodRequired())
+                        || (!$ctrl.model.defaultPaymentMethod && $ctrl.eligibility.isDefaultPaymentMethodChoiceRequired())
                         || $ctrl.model.challenge.checking
-                        || $ctrl.globalLoading.finalize)"
-                    data-ng-if="$ctrl.defaultPaymentMethod || !$ctrl.model.paymentMethod"
+                        || $ctrl.globalLoading.finalize
+                        || $ctrl.globalLoading.setDefaultPaymentMehtod"
+                    data-ng-if="$ctrl.defaultPaymentMethod || !$ctrl.model.paymentMethod || ($ctrl.eligibility.isDefaultPaymentMethodChoiceRequired() && !$ctrl.model.paymentMethod)"
```
Comment (cbourgois) [inline packages/manager/modules/pci/src/projects/new/payment/index.html:89]: (hard to read)

File: packages/manager/modules/pci/src/projects/new/payment/controller.js (hunk 11 above)
Comment (Jisay) [reply packages/manager/modules/pci/src/projects/new/payment/controller.js:170]: Not here, but when the promises are created. 
Note that I have to add it for set as default payment method promise cause I didn't add it ^^

File: packages/manager/modules/pci/src/projects/new/payment/components/add/index.html (hunk 7 above)
Comment (Jisay) [reply packages/manager/modules/pci/src/projects/new/payment/components/add/index.html:4]: I'm not sure to understand: you mean that there is no need to have this `ng-repeat` to create rows for the cols?
I think that a `div.row` parent is required for `div.col-*`

File: packages/manager/modules/pci/src/projects/new/payment/controller.js (hunk 11 above)
Comment (Jisay) [reply packages/manager/modules/pci/src/projects/new/payment/controller.js:170]: Not here. Because `manageProjectCreation` is called in `then` callback and this method is handled with `$q`

File: packages/manager/modules/pci/src/projects/new/payment/routing.js (hunk 12 above)
Comment (Jisay) [reply packages/manager/modules/pci/src/projects/new/payment/routing.js:54]: To know if there is a default payment method, we have to get all payment methods. So here I don't use `getDefaultPaymentMethod` directly because I need to list payment methods. So I get all payment methods to first know if there is or not a default payment method and if not display a list of valid payment methods. But in all the cases we have to get all payment methods to have their details...

File: packages/manager/modules/pci/src/projects/new/payment/components/add/index.html (hunk 8 above)
Comment (Jisay) [reply packages/manager/modules/pci/src/projects/new/payment/components/add/index.html:27]: This is originally a simple move of the code in different files but yes you are right this is more readable and easier

File: packages/manager/modules/pci/src/projects/new/payment/index.html (hunk 14 above)
Comment (Jisay) [reply packages/manager/modules/pci/src/projects/new/payment/index.html:89]: Yes, hard to not forget a case too :-D

File: packages/manager/modules/pci/src/projects/new/payment/components/choose/index.scss (hunk 5 above)
Comment (Jisay) [reply packages/manager/modules/pci/src/projects/new/payment/components/choose/index.scss:1]: That should be the rule indeed and I prefer this approach too

File: packages/components/ng-ovh-payment-method/src/payment-method-type.class.js [hunk 15]
```diff
@@ -96,4 +99,42 @@ export default class OvhPaymentMethodType {
...
     ].includes(this.integration);
   }
+
+  /**
+   * Determine if payment method type is a Bank Account.
+   *
+   * @return {Boolean}
+   */
+  isBankAccount() {
+    return this.paymentType === PAYMENT_METHOD_TYPE_ENUM.BANK_ACCOUNT;
```
Comment (marie-j) [inline packages/components/ng-ovh-payment-method/src/payment-method-type.class.js:109]: Why do this needs to be duplicate ? Isn't there a parent class ?

File: packages/manager/modules/pci/src/projects/new/payment/components/add/component.js [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+
+export default {
+  name: 'pciProjectNewPaymentMethodAdd',
+  controller,
+  template,
+  bindings: {
+    eligibility: '<',
+    model: '<',
+    registerablePaymentMethods: '<',
+    globalLoading: '<',
```
Comment (ganeshkumar1989) [inline packages/manager/modules/pci/src/projects/new/payment/components/add/component.js:12]: Would be great if the dependencies are sorted :)

File: packages/components/ng-ovh-payment-method/src/payment-method-type.class.js (hunk 15 above)
Comment (Jisay) [reply packages/components/ng-ovh-payment-method/src/payment-method-type.class.js:109]: Mmmmh, can be done but we talk about two different concept here:
- payment methods which have payment method type ;
- payment method types which are used to add a new payment method of given type.

So I don't think it's a good solution to have a mix of concepts. 
Maybe a static method to centralize it or maybe define `paymentType` attribute of `OvhPaymentMethod` instance as an instance of `OvhPaymentMethodType`.

File: packages/manager/modules/pci/src/projects/new/payment/components/add/index.html (hunk 7 above)
Comment (marie-j) [reply packages/manager/modules/pci/src/projects/new/payment/components/add/index.html:4]: What I meant is that it might not be needed to split the initial array into chunks of 2 arrays each

File: packages/components/ng-ovh-payment-method/src/payment-method-type.class.js (hunk 15 above)
Comment (marie-j) [reply packages/components/ng-ovh-payment-method/src/payment-method-type.class.js:109]: Yes Maybe `OvhPaymentMethod.type` should be an instance of OvhPaymentMethodType

File: packages/components/ng-ovh-payment-method/src/available-payment-method.class.js [hunk 17]
```diff
@@ -0,0 +1,108 @@
+import { isNull } from 'lodash-es';
```
Comment (Jisay) [inline packages/components/ng-ovh-payment-method/src/available-payment-method.class.js:1]: This file is just a move from previous `payment-method-type.class.js` file

File: packages/components/ng-ovh-contacts/LICENSE [hunk 18]
```diff
@@ -1,29 +0,0 @@
-BSD 3-Clause License
```
Comment (marie-j) [inline packages/components/ng-ovh-contacts/LICENSE:1]: This change doesn't look related to the PR's purpose

File: packages/components/ng-ovh-payment-method/src/legacy/mean/payment-mean-type.class.js [hunk 19]
```diff
@@ -1,15 +1,15 @@
...
 
 export default class OvhPaymentMeanType {
   constructor(options = {}) {
     this.meanType = options.value;
     this.registerable = options.registerable;
   }
 
-  toPaymentMethodType() {
-    return new OvhPaymentMethodType({
+  toAvailablePaymentMethod() {
```
Comment (marie-j) [inline packages/components/ng-ovh-payment-method/src/legacy/mean/payment-mean-type.class.js:11]: I didn't see any `BREAKING CHANGE` comment but this change looks like one as package is published

Comment (Jisay) [reply packages/components/ng-ovh-payment-method/src/legacy/mean/payment-mean-type.class.js:18]: set `toPaymentMethodType` to deprecated to avoid breaking change.

**Comment by antleblanc** [comment]
Pull Request has been rebase on `develop` in order to be able to resolve existing conflicts!
Also the git commit history has been rewritten to remove some extra commits and the translation has been retrieved.

I'm waiting the status of all check in order to be able to merge it on `develop`.

File: packages/components/ng-ovh-contacts/LICENSE (hunk 18 above)
Comment (antleblanc) [reply packages/components/ng-ovh-contacts/LICENSE]: Thanks for pointing out this. It has been resolved via the rebase.
